File: graph/ruby/adjmat_w.rb
class Graph
    def initialize(nv, edges)
        @nvertices = nv
        @adjmat = nv.times.collect { Array.new(nv, 0) }
        edges.each do |i, j, w|
            @adjmat[i][j] = w
        end
    end
    def each_vertices
        @nvertices.times do |i|
            yield i
        end
    end
    def each_edges
        @nvertices.times do |i|
            @nvertices.times do |j|
                w = @adjmat[i][j]
                yield [i,j,w] if w != 0
            end
        end
    end
    def each_successors(i)
        @nvertices.times do |j|
            w = @adjmat[i][j]
            yield [j,w] if w != 0
        end
    end
    def vertices
        Enumerator.new(self, :each_vertices)
    end
    def edges
        Enumerator.new(self, :each_edges)
    end
    def successors(i)
        Enumerator.new(self, :each_successors, i)
    end
end

File: graph/haskell/AdjmatW.hs
module AdjmatW (
    Graph, Vertex, Edge,
    create, vertices, edges, successors
    ) where

import Data.Array

type Vertex  = Int
type Edge w  = (Vertex, Vertex, w)
type Graph w = Array (Vertex, Vertex) w

create :: Num w => Int -> [Edge w] -> Graph w
create nv edges = accumArray (\_ w -> w) 0
    ((0,0), (nv-1,nv-1)) [((i,j),w) | (i,j,w) <- edges]

nvertices :: Num w => Graph w -> Int
nvertices g = n + 1
    where
    (_, (n, _)) = bounds g

vertices :: Num w => Graph w -> [Vertex]
vertices g = [0..nvertices g]

edges :: Num w => Graph w -> [Edge w]
edges g = [ (i,j,w) | ((i,j), w) <- assocs g, w /= 0]

successors :: Num w => Graph w -> Vertex -> [(Vertex, w)]
successors g i = [ (j, w) | j <- [0..n], let w = g!(i, j), w /= 0]
    where
    (_, (n, _)) = bounds g


